### ISSUE TITLE
e2e does not set minimum cluster size when creating cluster

### ISSUE THREAD

**Issue by BenTheElder** [GoogleCloudPlatform/kubernetes#7687]
While running e2e on vagrant I discovered i needed to run `export NUM_MINIONS=2` before running the tests or else get the `Cluster size (1) is too small to run e2e tests.  2 Minions are required.` error.

In my use case I'm working on the proxy code, so I do not have a cluster up and i call something like 'hack/e2e-test.sh' to put up a cluster and run the end to end tests on the new code which does:
`go run hack/e2e.go-build -up -test -down`
Which in turn calls cluster/vagrant/util.sh which imports the vagrant default environment settings, either re-exporting the current NUM_MINIONS or 1, when for the tests we need NUM_MINIONS >= 2.

I think the ideal behavior would check NUM_MINONS and adjust accordingly, or at least some documentation noting that this needs to be set correctly (here?: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/development.md#end-to-end-tests).

**Comment by roberthbailey** [comment]
It seems like adding a quick check in the e2e test code wouldn't be too difficult so that we would at least fail more quickly.

**Comment by BenTheElder** [comment]
That's what I thought, but if -up is set shouldn't it create a cluster of the right size for the tests?

**Comment by roberthbailey** [comment]
It just calls `kube-up` and assumes that the provider has the correct configuration. I'm not sure that there is an easy way to check that NUM_MINIONS is set properly until after the cluster is created though given the mess of Go/bash interdependency here...

The e2e test driver sets `KUBE_CONFIG_FILE` to 'config-test.sh` so that it's easy to specify a different default configuration for testing than for normal cluster creation (see https://github.com/GoogleCloudPlatform/kubernetes/blob/master/hack/e2e.go#L151) so it should also be easy to fix vagrant's config-test.sh file to specify that there should be 2 minions.

**Comment by BenTheElder** [comment]
That sounds like a good fix. Should this be done just for vagrant or for all of the providers?

**Comment by BenTheElder** [comment]
What about something like:
````shell
if [[ NUM_MINONS -lt 2 ]]; then
    echo "NUM_MINIONS < 2 ; 2 or more minions are required. Setting NUM_MINONS=2"
    NUM_MINONS=2
fi
````

**Comment by roberthbailey** [comment]
You could probably add that to https://github.com/GoogleCloudPlatform/kubernetes/blob/master/hack/e2e-internal/e2e-up.sh although I think it would be better to fail rather than change the number of minions. What if a cloud provider doesn't support two minions (e.g. the single node running in docker case)?

**Comment by BenTheElder** [comment]
Good point.
So then just the config-test.sh for vagrant should have `NUM_MINIONS=2` and maybe the quick check in e2e? I can make a PR for that.

**Comment by roberthbailey** [comment]
That'd be great. Thanks Ben!

@derekwaynecarr as an FYI for the vagrant changes.

**Comment by BenTheElder** [comment]
You're welcome! Thanks for the guidance; I'm just happy to be working on a interesting project with a friendly community.
I'm wrapping up the other pr for the changes to e2e right now, just running the tests.

**Comment by BenTheElder** [comment]
Turns out the cluster size is validated _after_ the cluster is created because there isn't any clean way to get NUM_MINIONS from the environment and config scripts.
I think just changing the default NUM_MINIONS for vagrant tests solves this for now for me, but i'll come back to it if I think of something.

In the meantime I did find a comment in e2e talking about a minimum of 3 minions (it should be 2) that i'll submit a PR fixing.
